frontend/src/views/admin/matching/reducers.ts

```diff
@@ -154,4 +156,114 @@ const matchingDataReducer = createReducer(initialState, {
     [SET_UPDATED]: (state, action) => {
         return { ...state, updated: action.payload };
     },
+    [TOGGLE_STARRED]: (state, action) => {
+        // Check if a match with this applicant ID and position ID already exists
+        const existingMatch = state.matches.find(
+            (match) =>
+                match.utorid === action.payload.utorid &&
+                match.positionCode === action.payload.positionCode
+        );
+
+        // If a match doesn't already exist, this applicant is implied to not yet be starred:
+        if (!existingMatch) {
+            const newMatch: RawMatch = {
+                utorid: action.payload.utorid,
+                positionCode: action.payload.positionCode,
+                starred: true,
+            };
+
+            return {
+                ...state,
+                matches: [...state.matches, newMatch],
+                updated: true,
+            };
+        }
+
+        // Item exists, so we have to update it
+        return {
+            ...state,
+            matches: state.matches
+                .map((match) => {
+                    if (
+                        match.utorid === action.payload.utorid &&
+                        match.positionCode === action.payload.positionCode
+                    ) {
+                        // Check if the match will have any flags set, otherwise mark for deletion:
+                        if (
+                            match.stagedAssigned ||
+                            !match.starred ||
+                            match.hidden
+                        ) {
+                            return { ...match, starred: !match.starred };
+                        }
+                        return null;
+                    } else {
+                        return match;
+                    }
+                })
+                .filter((match): match is RawMatch => !!match),
+            updated: true,
+        };
+    },
+    [TOGGLE_ASSIGNED]: (state, action) => {
+        // Check if a match with this applicant ID and position ID already exists
+        const existingMatch = state.matches.find(
+            (match) =>
+                match.utorid === action.payload.utorid &&
+                match.positionCode === action.payload.positionCode
+        );
+
+        // If a match doesn't already exist, this applicant is implied to not yet be assigned:
+        if (!existingMatch) {
+            const newMatch: RawMatch = {
+                utorid: action.payload.utorid,
+                positionCode: action.payload.positionCode,
+                stagedHoursAssigned: action.payload.stagedHoursAssigned || 0,
+                stagedAssigned: true,
+                starred: false,
+                hidden: false,
+            };
+
+            return {
+                ...state,
+                matches: [...state.matches, newMatch],
+                updated: true,
+            };
+        }
+
+        return {
```

Comment:
Code like this makes me wonder if we should just start using the `immer` library...Do you have experience with it? Did you want to investigate and see if it's worth the dependency?